### PR TITLE
lending: Use fair obligation health factor calculation

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
Loans that are large in value relative to the depth of the dex order book are prone to having their collateral devalued which leads to a higher LTV ratio and potential opportunity for liquidations.

#### Changes
- Use the best price in the order book to determine the health of an obligation but continue using order book depth to calculate a fair market value of of collateral to send to liquidators. This leaves large obligations much less vulnerable to order book manipulation and provides an efficient way to determine the health of a loan. An additional win is that the amount of liquidity a liquidator wants to pay off has no impact on the determination about an obligation's health.
- Add special handling for liquidating dust obligations. When an obligation has less than a single token unit of borrowed value remaining, allow a liquidator to pay it off in return for the remaining collateral.

### CODE REVIEW

File: token-lending/program/src/state/reserve.rs [hunk 1]
```diff
@@ -95,41 +95,46 @@ impl Reserve {
...
         let max_repayable = obligation
             .borrowed_liquidity_wads
             .try_mul(Rate::from_percent(LIQUIDATION_CLOSE_FACTOR))?
             .try_floor_u64()?;
-        let integer_repay_amount = liquidity_amount.min(max_repayable);
-        if integer_repay_amount == 0 {
-            return Err(LendingError::ObligationTooSmall.into());
+
+        // Special handling for dust obligations
+        if max_repayable == 0 {
```
Comment (joncinque) [inline token-lending/program/src/state/reserve.rs:121]: Were you able to check that the proptests hit this case?  If not, could you add a special test case for this?

**Review by joncinque — APPROVED** [review]
Looks good!  Just two small questions really, which don't need to be done in this PR.

File: token-lending/program/src/state/reserve.rs (hunk 1 above)
Comment (joncinque) [reply token-lending/program/src/state/reserve.rs:121]: Also, can this happen in normal repays too?

Comment (jstarry) [reply token-lending/program/src/state/reserve.rs:121]: Yes, there's already handling for normal repays for this case. I'll add some explicit test cases

**Comment by jstarry** [comment]
@joncinque do you mind taking a look at the latest changes? In writing your requested tests I found some holes.
- Repay amount for dust accounts was hard coded to `1` but should be the ceiling of the decimal borrow amount
- If decimal borrow amount was between 1 and 2, we weren't enforcing that the input liquidity amount was greater than 1.
- There were some edge cases where unhealthy loans couldn't be liquidated, I fixed those and refactored the test to ensure that all unhealthy loans can be liquidated.
- Decided that splitting off the bonus amount wasn't worth the compute cost and by combining the bonus and withdraw amounts, we're now giving liquidators a more fair amount of collateral (rather than rounding down separately)

File: token-lending/program/src/dex_market.rs [hunk 2]
```diff
@@ -59,6 +59,43 @@ pub struct TradeSimulator<'a> {
...
+
+        let order_book_side = match (action, currency) {
+            (TradeAction::Buy, Currency::Base) => Side::Ask,
+            (TradeAction::Sell, Currency::Quote) => Side::Ask,
+            (TradeAction::Buy, Currency::Quote) => Side::Bid,
+            (TradeAction::Sell, Currency::Base) => Side::Bid,
+        };
+        if order_book_side != self.orders_side {
+            return Err(LendingError::DexInvalidOrderBookSide.into());
+        }
```
[the remark this answers is not in the document]
Comment (jstarry) [reply token-lending/program/src/dex_market.rs:83]: will do later

**Comment by joncinque** [comment]
Looks really good, just some questions and nits

File: token-lending/program/src/state/reserve.rs [hunk 3]
```diff
@@ -795,84 +788,102 @@ mod test {
...
-        fn liquidate_obligation(
-            obligation_loan in 1..=u64::MAX,
-            obligation_collateral in 1..=u64::MAX,
-            liquidate_amount in 1..=u64::MAX,
-            collateral_exchange_rate in 1..=500u64,
-            token_conversion_rate in 1..=u64::MAX,
-            liquidation_bonus in 0..=100u8,
-            liquidation_threshold in 2..=100u8,
+        fn unhealthy_obligations_can_be_liquidated(
+            obligation_collateral in 1..=u32::MAX,
```
[the remark this answers is not in the document]
Comment (jstarry) [reply token-lending/program/src/state/reserve.rs:826]: Yes, fixed

File: token-lending/program/src/state/reserve.rs [hunk 4]
```diff
@@ -92,85 +95,90 @@ impl Reserve {
...
-
-        // Don't pay out bonus if withdraw amount covers full collateral balance
-        let (withdraw_amount, bonus_amount, remaining_amount) =
-            if withdraw_amount >= obligation.deposited_collateral_tokens.into() {
-                (obligation.deposited_collateral_tokens, 0, 0)
+        // Calculate the amount of liquidity that will be repaid
+        let max_repay_amount = obligation
+            .borrowed_liquidity_wads
+            .try_mul(Rate::from_percent(LIQUIDATION_CLOSE_FACTOR))?
+            .try_floor_u64()?;
```
[the remark this answers is not in the document]
Comment (jstarry) [reply token-lending/program/src/state/reserve.rs]: Yes, not sure how that happened.

**Review by jstarry — COMMENTED** [review]
Thanks for the detailed feedback @joncinque!

**Comment by jstarry** [comment]
feedback changes can be found here: https://github.com/solana-labs/solana-program-library/commit/ca5e43f3c28b205a8c10342d47ce5a33a75de925

File: token-lending/program/src/state/reserve.rs [hunk 5]
```diff
@@ -92,85 +95,90 @@ impl Reserve {
...
-        // Calculate the amount of liquidity that will be repaid
-        let max_repayable = obligation
-            .borrowed_liquidity_wads
-            .try_mul(Rate::from_percent(LIQUIDATION_CLOSE_FACTOR))?
-            .try_floor_u64()?;
-
-        // Special handling for dust obligations
-        if max_repayable == 0 {
+        // Special handling for small, closeable obligations
+        let close_amount = liquidity_amount.min(CLOSEABLE_AMOUNT);
```
[the remark this answers is not in the document]
Comment (joncinque) [reply token-lending/program/src/state/reserve.rs]: Right, of course, I forgot about the 50% max liquidation!  Ok great, thanks for the explanation

**Comment by joncinque** [comment]
Looks good
